src/otx/cli/cli.py

```diff
@@ -0,0 +1,325 @@
+
+
+class OTXCLI:
+    """OTX CLI entrypoint."""
+
+    def __init__(self) -> None:
+        """Initialize OTX CLI."""
+        self.console = Console()
+        self._subcommand_method_arguments: dict[str, list[str]] = {}
+        with patch_update_configs():
+            self.parser = self.init_parser()
+            self.add_subcommands()
+            self.config = self.parser.parse_args(_skip_check=True)
+
+        self.subcommand = self.config["subcommand"]
+        self.run()
+
+    def init_parser(self) -> ArgumentParser:
+        """Initialize the argument parser for the OTX CLI.
+
+        Returns:
+            ArgumentParser: The initialized argument parser.
+        """
+        parser = ArgumentParser(
+            description="OpenVINO Training-Extension command line tool",
+            env_prefix="otx",
+            formatter_class=CustomHelpFormatter,
+        )
+        parser.add_argument(
+            "-v",
+            "--version",
+            action="version",
+            version=f"%(prog)s {__version__}",
+            help="Display OTX version number.",
+        )
+        return parser
+
+    def subcommand_parser(self, **kwargs) -> ArgumentParser:
+        """Returns an ArgumentParser object for parsing command line arguments specific to a subcommand.
+
+        Returns:
+            ArgumentParser: An ArgumentParser object configured with the specified arguments.
+        """
+        parser = ArgumentParser(
+            formatter_class=CustomHelpFormatter,
+            **kwargs,
+        )
+        parser.add_argument(
+            "-v",
+            "--verbose",
+            action="count",
+            help="Verbose mode. This shows a configuration argument that allows for more specific overrides. \
+                Multiple -v options increase the verbosity. The maximum is 2.",
+        )
+        parser.add_argument(
+            "-c",
+            "--config",
+            action=ActionConfigFile,
+            help="Path to a configuration file in json or yaml format.",
+        )
+        parser.add_argument(
+            "--data_root",
+            type=str,
+            help="Path to dataset root.",
+        )
+        parser.add_argument(
+            "--task",
+            type=str,
+            help="Task Type.",
+        )
+        parser.add_argument(
+            "--callback_monitor",
+            type=str,
+            help="The metric to monitor the model performance during training callbacks.",
+        )
+        return parser
+
+    @staticmethod
+    def engine_subcommands() -> dict[str, set[str]]:
+        """Returns a dictionary of engine subcommands and their required arguments.
+
+        Returns:
+            A dictionary where the keys are the subcommands and the values are sets of required arguments.
+        """
+        device_kwargs = {"accelerator", "devices"}
+        return {
+            "train": device_kwargs,
+            "test": {"datamodule"}.union(device_kwargs),
```

Comment:
Why we need {"datamodule"}.union?

Reply:
In the CLI flow, the `datamodule` comes in when the engine is created. However, engine can also get a `datamodule` in `test()`. This is a redundant case, so skip it in the CLI to disallow it.